Heuristics drift time

willquinn opened the pull request:
Features added:
- HPGe scalar field (LH5) specification and routines to read them in memory
- HPGe drift time processor, using `ak.transform()`
- drift-time heuristic processor, based on Awkward arrays and sped up with Numba
- new `units.py` module with first routines to handle physical units
- unit tests

Outstanding tasks:
- full profiling needs to be done to test speed
- validate distributions on high-statistics test data

willquinn commented:
I will add doc strings to all these functions.

tdixon97 commented on src/reboost/hpge/drift_time.py:
```diff
@@ -0,0 +1,86 @@
+        self.t = t
+
+
+class ReadDTFile:
```
I would like to move the more generic methods into the `utils` eg. interpolation.

willquinn replied:
I'll move these classes to utils.py

tdixon97 commented on src/reboost/hpge/drift_time.py:
```diff
@@ -0,0 +1,86 @@
+
+
+class DriftTimes:
+    def __init__(self, r, z, t) -> None:
```
All functions need type hints.

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +75,89 @@ def cumsum(layout, **_kwargs):
     return Array(ak.flatten(r90).to_numpy())
+
+
+def calculate_drift_times(
```
is this a processor ? We should make clear whats part of the public API and what isnt

willquinn replied:
Should I move this to a dedicated file -> drift_times.py?

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +75,89 @@ def cumsum(layout, **_kwargs):
+
+
+@numba.njit(cache=True)
+def psa(t1: np.float64, e1: np.float64, t2: np.float64, e2: np.float64) -> np.float64:
```
maybe needs _ ?

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +75,89 @@ def cumsum(layout, **_kwargs):
+def e_scaler(e1: np.float64, e2: np.float64) -> np.float64:
+    if e1 > 0 and e2 > 0:
+        return 1 / np.sqrt(e1 * e2)
+    return 1e9  # Large value to avoid divide-by-zero errors
```
There has got to be a better solution than this

willquinn replied:
I think I deal with it higher up the chain, so I can get rid of this an change to `return 1 / np.sqrt(e1 * e2)`

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +75,89 @@ def cumsum(layout, **_kwargs):
+
+
+def dt_heuristic(data: ak.Array, dt_file: ReadDTFile) -> Array:
+    drift_times = calculate_drift_times(data.xloc, data.yloc, data.zloc, dt_file)
```
is `dt_file` an object containing the dt map, if so we need a better name.

tdixon97 commented on tests/hpge/test_dt_heuristic.py:
```diff
@@ -0,0 +1,99 @@
+    }
+    clusters["edep"] = cluster_energy
+
+    clusters["cluster_size"] = (
```
this calculations are rather long, I suggest a processor

willquinn replied:
Shall I create a new cluster processor and add it to cluster.py?

tdixon97 commented on tests/hpge/test_dt_heuristic.py:
```diff
@@ -0,0 +1,99 @@
+    return ak.Array(clusters)
+
+
+def test_dt_heuristic():
```
this only tests it runs, which is ok. But what about some tests of some simple cases

willquinn replied:
Yes I'll add in some assertions on the output

tdixon97 commented:
I think it needs alot more documentation, and we also need to think about what parts of this are general and which are specific. I want to build some basic functionality to allow people to do similar things in future

willquinn commented:
I can use the cases I defined in r90 and do some more simple tests, though the test_files have 10 events in so it is simple. I just need to check what I expect the output to be.

tdixon97 commented:
I suggest some more trivial cases where the output is known (eg. uniform drift time), one step etc etc.

willquinn commented:
Regarding the MPP clustering, I checked your function and compared it to the one in MPP, and it seemed the same. The way I interpreted the algorithm in cluster.py is that you check distances and track ids. I set all steps in the event to have the same track ID by using the event ID.

tdixon97 commented:
But does MPP cluster together steps in different tracks?

willquinn commented:
Here is the MPP clustering code - modified slightly by myself, but the main function is the same:
```
void MPPStepsClusterer::SensVolProcess()
{
  // First figure out what order we want to go through fSteps and fActiveness in. We want to go through the steps in numerical order so that tracks will be grouped together and in order
  fNSensVol++;
  vector<size_t> i_order(fSteps->size());
  iota(i_order.begin(), i_order.end(), 0);
  sort(i_order.begin(), i_order.end(), [this](size_t lhs, size_t rhs) { return make_pair((*fSteps)[lhs]->GetTrackID(), (*fSteps)[lhs]->GetStepNumber()) < make_pair((*fSteps)[rhs]->GetTrackID(), (*fSteps)[rhs]->GetStepNumber()); });
  
  fClusters.clear();
  fNClusters=0;
  fNSteps.clear();
  fClusterSteps.clear();
  fClusterEnergy.clear();
  fClusterActiveness.clear();
  fClusterSize.clear();
  fClusterPositionX.clear();
  fClusterPositionY.clear();
  fClusterPositionZ.clear();
  
  const MGTMCStepData* laststep = NULL;
  size_t icluster = 0;
  vector<vector<size_t> > cluster_steps;
  cluster_steps.reserve(fSteps->size());
  
  // Loop through the steps and add them to the correct cluster
  for(size_t i : i_order) {
    const MGTMCStepData* thisstep = (*fSteps)[i];

    // If this is a prestep, seek out the cluster/step that spawned it
    if(thisstep->GetIsPreStep()) {
      laststep = NULL;
      // loop through each step in each cluster until we find the parent
      for(size_t jcluster=0; jcluster<cluster_steps.size(); jcluster++) {
        for(size_t iStep : cluster_steps[jcluster]) {
          const MGTMCStepData* step = (*fSteps)[iStep];
          if(thisstep->GetParentTrackID() == step->GetTrackID()
            && TMath::Abs(thisstep->GetT() - step->GetT()) < fClusterTime
            && (thisstep->GetPositionVector()-step->GetPositionVector()).Mag() < fClusterDistance) {
            icluster = jcluster;
            laststep = step;
            break;
          }
        }
	      if(laststep) break;
      }

      // do not place presteps in clusters
      continue;
    }
    
    // if this step is not within the cluster distance/time of the last step (or there is no last step), start a new cluster
    if(!(laststep
	 && ( thisstep->GetTrackID() == laststep->GetTrackID()
	      || thisstep->GetParentTrackID() == laststep->GetTrackID() )
	 && thisstep->GetStepLength() < fClusterDistance
	 && TMath::Abs(laststep->GetT()-thisstep->GetT()) < fClusterTime) ) {
      cluster_steps.emplace_back();
      cluster_steps.back().reserve(fSteps->size());
      icluster = cluster_steps.size()-1;
      fNClusters++;
    }
    // Add this step to the cluster if it is not a prestep. All non-pre-steps are guaranteed to be in a cluster
    cluster_steps[icluster].push_back(i);
    fNTotSteps++;
    laststep = thisstep;
  }
  
  // now make clusters that are averages of lists of steps
  for(size_t iClust=0; iClust<cluster_steps.size(); iClust++) {
    vector<size_t>& steps = cluster_steps[iClust];
    
    // Get new cluster from cache (or create new one)
    if(fClusterIt==fClusterCache.end())
      fClusterIt = fClusterCache.emplace(fClusterIt);
    MGTMCStepData& clust = *fClusterIt;
    fClusterIt++;
    // initialize cluster with first step
    clust = *((*fSteps)[steps[0]]);
    clust.SetEdep(0.);
    fClusters.push_back(&clust);
    fClusterSteps.emplace_back(0);
    fClusterEnergy.push_back(0.);
    fClusterActiveness.push_back(0.);
    fClusterSize.push_back(0.);
    fClusterPositionX.push_back(0.);
    fClusterPositionY.push_back(0.);
    fClusterPositionZ.push_back(0.);
    
    // loop through other steps and average them in
    for(size_t iStep : steps) {
      const MGTMCStepData* step = (*fSteps)[iStep];
      fClusterSteps[iClust].push_back(step);
      
      double activeness = ( fActiveness!=NULL ? (*fActiveness)[iStep] : 1. );
      clust.SetEdep( clust.GetEdep()+step->GetEdep() );
      double w = ( clust.GetEdep()==0 ? 1. : step->GetEdep()/clust.GetEdep());
      
      fClusterEnergy[iClust] += step->GetEdep()*activeness;
      fClusterActiveness[iClust] += w*(activeness - fClusterActiveness[iClust]);
      fClusterSize[iClust] += (1.-w) * ( (step->GetPositionVector()-clust.GetPositionVector()).Mag2() - fClusterSize[iClust] );
      
      clust.SetT( w*step->GetT() + (1.-w)*clust.GetT());
      clust.SetX( w*step->GetX() + (1.-w)*clust.GetX());
      clust.SetY( w*step->GetY() + (1.-w)*clust.GetY());
      clust.SetZ( w*step->GetZ() + (1.-w)*clust.GetZ());
      clust.SetLocalX( w*step->GetLocalX() + (1.-w)*clust.GetLocalX());
      clust.SetLocalY( w*step->GetLocalY() + (1.-w)*clust.GetLocalY());
      clust.SetLocalZ( w*step->GetLocalZ() + (1.-w)*clust.GetLocalZ());
    }
    if(fWindowT0) clust.SetTimeOffset(clust.GetT() - *fWindowT0);
    fNSteps.push_back(fClusterSteps[iClust].size());
    fClusterSize[iClust] = sqrt(fClusterSize[iClust]);
    fClusterPositionX[iClust] = clust.GetLocalX();
    fClusterPositionY[iClust] = clust.GetLocalY();
    fClusterPositionZ[iClust] = clust.GetLocalZ();
  }

  fTotalNClusters += fNClusters;
  if(GetVerbosity() > 0) {
    for(size_t iC=0; iC<fClusters.size(); iC++) {
      const MGTMCStepData* clust = fClusters[iC];
      printf("  %4lu  %8d  %5lu  %6lu  %8.2e  (%6.2f, %6.2f, %6.2f)  %7.2f  %5.3f  %8.2e\n",
	     IWindow(), SensVolID(), iC, fNSteps[iC],
	     clust->GetTimeOffset(), clust->GetLocalX(), clust->GetLocalY(),
	     clust->GetLocalZ(),  clust->GetEdep()*1000,
	     fClusterActiveness[iC], fClusterSize[iC]);
    }
  }
}
```

gipert commented:
hi willie what's the status here?

willquinn commented:
I have restructured some of the code, added some doc strings.

My implemention of some clustering needs some work, also not sure if its in the right place.

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+    return 1 / np.sqrt(e1 * e2)
+
+
+def do_cluster(grouped_data: ak.Array, cluster_size_mm: np.float64):
```
this function is unnecessary, we can just use a sequence of processors

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+
+
+@numba.njit(cache=True)
+def identification_metric(
```
are these functions of general use, if not maybe mark them as private with a _ or define them inside the function using them? I would rather not clutter the API  so much

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+        f"{field}": apply_cluster(cluster_indices, grouped_data[f"{field}"]).view_as("ak")
+        for field in ["edep", "xloc", "yloc", "zloc", "time", "dist_to_surf"]
+    }
+    clustered_data["activeness"] = piecewise_linear_activeness(
```
we should not be hardcoding a particular activeness function

willquinn replied:
Where should be this controlled, when calling the dt heuristic?

tdixon97 commented on src/reboost/hpge/drift_time.py:
```diff
@@ -0,0 +1,25 @@
+def calculate_drift_times(
+    xdata: ak.Array, ydata: ak.Array, zdata: ak.Array, drift_time_map: ReadHPGeMap
+) -> ak.Array:
+    x_flat = ak.flatten(xdata).to_numpy()
```
can it not be made more elegant?
eg 
``` position_flat =[ ak.flatten(axis_data).to_numpy() for axis_data in [xdata,ydata,zdata] ]```

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+    """
+    cluster_indices = cluster_by_step_length(
+        grouped_data["evtid"],
+        grouped_data["xloc"] * 1000,  # has to be in mm
```
I think we have to handle units better (here and everywhere), this is very unsafe

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+    }
+    clusters["edep"] = cluster_energy
+
+    clusters["cluster_size"] = (
```
reduce the number of lines of code here

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +78,234 @@ def cumsum(layout, **_kwargs):
+    - Then`calculate_dt_heuristic()` is called which does the computation.
+    """
+    if "activeness" not in data.fields:
+        activeness = piecewise_linear_activeness(
```
lets not hardcode the activeness function, maybe try to think about how this can be done with sequential processors

tdixon97 commented on src/reboost/hpge/utils.py:
```diff
@@ -0,0 +1,89 @@
+from pyg4ometry.geant4 import LogicalVolume
+
+
+class HPGeMapPoint:
```
this does not need to be a class

willquinn replied:
ok I will change this

tdixon97 commented on src/reboost/hpge/utils.py:
```diff
@@ -0,0 +1,89 @@
+        self.val = val
+
+
+class ReadHPGeMap:
```
I don't like at all the name of this class (and I'm dubious if it really needs to be).
It would be good if you can also make this generic for maps depending on another property (eg energy)

willquinn replied:
without seeing other maps I don't know how to make this more generic. I assumed that I would have a column for z, a column for r and a column for the value of the field (energy, A/E, drift time etc). This class should work for any of them. I don't see why we can't use a class to be honest.

tdixon97 replied:
I would write something generic based on N, explanatory variables ($x_i$) and 1 target variable ($y$) that does an N-d interpolation predicting the output field. This could be applied to eg.

```
(r,z) --> dt
(x,y,z) -> dt
(r,z,E) --> A/E
(r, z, E1, E2)--> A/E
```
whatever you want...
I think the data can be held in a standard python data structure (such as a dict or named tuple) and you just need a method computing:
```python
interpolate(data:AttrsDict,parameters:NamedTuple)
```
or something similar. If you really want to make it a class thats ok, but for something like this its a bit unnecessary. See this page [link](https://learn.scientific-python.org/development/principles/design/#consider-can-this-just-be-a-function)

tdixon97 commented on src/reboost/hpge/utils.py:
```diff
@@ -0,0 +1,89 @@
+            return nDT - 1
+        return i
+
+    def get_map_value(self, x: np.float64, y: np.float64, input_z: np.float64) -> np.float64:
```
i think interpolate would be a better name

willquinn commented on src/reboost/hpge/utils.py:
```diff
@@ -0,0 +1,89 @@
+        self.minr = min(rs)
+
+    def get_map_index(self, r: np.float64, z: np.float64) -> int:
+        rGrid = floor(r / self.dr + 0.5) * self.dr
```
its a relic of the C++ code. Easy change

tdixon97 replied:
which C++ code..?

willquinn commented:
@gipert not sure why some many checks are failing

willquinn commented:
Maybe I should write it here to make it clear what I have assumed so far. 

I have assumed there would be 1 drift-time-map lh5 file generated for a given period of data (i.e for a given HPGe HV). This file will contain all detectors like so:
`
/
└── B99000A · table{r,z,dt} 
    ├── dt · array<1>{real} 
    ├── r · array<1>{real} 
    └── z · array<1>{real} 
`
So the read_hpge_map function produces a dictionary with all the information for a given HPGe. I tried to make this generic so you can format the drift map in terms of (r,z), (x,y - not sure why you would want this) or (x, y, z). But mainly this is interpreted in the interpolation functions, 2D or 3D. The difficulty is that the input will be x/y/zlocs with global coords but the HPGe map requires relative coords. So I have added this as a special case with a boolean. This part of the code is the part I need some opinions on as I don't know what is best.

I have changed the code so the calculation of drift times is its own processor.

The dt heuristic part is just a processor.

tdixon97 commented:
You just need to add processors before calling the drift time who compute local coordinates. This we can easily do
________________________________
From: William Quinn ***@***.***>
Sent: Wednesday, April 9, 2025 6:41 PM
To: legend-exp/reboost ***@***.***>
Cc: Dixon, Toby ***@***.***>; Review requested ***@***.***>
Subject: Re: [legend-exp/reboost] Heuristics drift time (PR #52)


⚠ Caution: External sender


Maybe I should write it here to make it clear what I have assumed so far.

I have assumed there would be 1 drift-time-map lh5 file generated for a given period of data (i.e for a given HPGe HV). This file will contain all detectors like so:

/
└── B99000A · table{r,z,dt}
├── dt · array<1>{real}
├── r · array<1>{real}
└── z · array<1>{real}

So the read_hpge_map function produces a dictionary with all the information for a given HPGe. I tried to make this generic so you can format the drift map in terms of (r,z), (x,y - not sure why you would want this) or (x, y, z). But mainly this is interpreted in the interpolation functions, 2D or 3D. The difficulty is that the input will be x/y/zlocs with global coords but the HPGe map requires relative coords. So I have added this as a special case with a boolean. This part of the code is the part I need some opinions on as I don't know what is best.

I have changed the code so the calculation of drift times is its own processor.

The dt heuristic part is just a processor.

—
Reply to this email directly, view it on GitHub<https://github.com/legend-exp/reboost/pull/52#issuecomment-2790355147>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANSET4YPSSKVVFHOWEBTGNT2YVEVLAVCNFSM6AAAAABXDQ46ZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJQGM2TKMJUG4>.
You are receiving this because your review was requested.Message ID: ***@***.***>

[https://avatars.githubusercontent.com/u/44215491?s=20&v=4]willquinn left a comment (legend-exp/reboost#52)<https://github.com/legend-exp/reboost/pull/52#issuecomment-2790355147>

Maybe I should write it here to make it clear what I have assumed so far.

I have assumed there would be 1 drift-time-map lh5 file generated for a given period of data (i.e for a given HPGe HV). This file will contain all detectors like so:

/
└── B99000A · table{r,z,dt}
├── dt · array<1>{real}
├── r · array<1>{real}
└── z · array<1>{real}

So the read_hpge_map function produces a dictionary with all the information for a given HPGe. I tried to make this generic so you can format the drift map in terms of (r,z), (x,y - not sure why you would want this) or (x, y, z). But mainly this is interpreted in the interpolation functions, 2D or 3D. The difficulty is that the input will be x/y/zlocs with global coords but the HPGe map requires relative coords. So I have added this as a special case with a boolean. This part of the code is the part I need some opinions on as I don't know what is best.

I have changed the code so the calculation of drift times is its own processor.

The dt heuristic part is just a processor.

—
Reply to this email directly, view it on GitHub<https://github.com/legend-exp/reboost/pull/52#issuecomment-2790355147>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANSET4YPSSKVVFHOWEBTGNT2YVEVLAVCNFSM6AAAAABXDQ46ZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOOJQGM2TKMJUG4>.
You are receiving this because your review was requested.Message ID: ***@***.***>

gipert commented:
@willquinn can you share the Julia code to produce the drift time maps?

tdixon97 commented:
What repo should this belong? Maybe Oli has some idea
________________________________
From: Luigi Pertoldi ***@***.***>
Sent: Wednesday, April 23, 2025 10:45 AM
To: legend-exp/reboost ***@***.***>
Cc: Dixon, Toby ***@***.***>; Review requested ***@***.***>
Subject: Re: [legend-exp/reboost] Heuristics drift time (PR #52)


⚠ Caution: External sender


@willquinn<https://github.com/willquinn> can you share the Julia code to produce the drift time maps?

—
Reply to this email directly, view it on GitHub<https://github.com/legend-exp/reboost/pull/52#issuecomment-2823525153>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANSET42DCLOG5QD2YRMFCEL225HKJAVCNFSM6AAAAABXDQ46ZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRTGUZDKMJVGM>.
You are receiving this because your review was requested.Message ID: ***@***.***>

[https://avatars.githubusercontent.com/u/20358192?s=20&v=4]gipert left a comment (legend-exp/reboost#52)<https://github.com/legend-exp/reboost/pull/52#issuecomment-2823525153>

@willquinn<https://github.com/willquinn> can you share the Julia code to produce the drift time maps?

—
Reply to this email directly, view it on GitHub<https://github.com/legend-exp/reboost/pull/52#issuecomment-2823525153>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ANSET42DCLOG5QD2YRMFCEL225HKJAVCNFSM6AAAAABXDQ46ZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRTGUZDKMJVGM>.
You are receiving this because your review was requested.Message ID: ***@***.***>

gipert commented:
in the long run it's supposed to be a short script based on LegendGeSim, i think we can store it in this repo

willquinn commented:
Summary of what I am up to:

- Adding the julia script for generating the drift map (now in correct format with 1d r and z and 2d dt)
- scipy interpolate function

The actual dt heuristic has been sorted for a while I think

willquinn commented:
Also the failures come from a division by zero warning - I will need to isolate this and deal with it

gipert commented:
I have already some changes locally also about your second bullet, can you work on the first for now?

willquinn commented:
The detectors are irregular shapes, there are gaps for the tapering, the bore hole and the grooves - should these just be nans in the 2d array? This depends on your interpolation... what I had was trying to sort this, so I guess just tell me what you want

gipert commented:
Ah I see the point. Why not having a rectangular grid that contains the detector profile and set the drift times to zero where there is no Ge?

tdixon97 commented:
whats the drift time in the dead-layer?

gipert commented:
whatever SSD determines, what's the problem?

gipert commented:
I added a specification for rectangular gridded scalar fields and methods to read/interpolate. please have a look. @willquinn if you could prepare a test DT map in this format I can go ahead with adapting the rest

tdixon97 commented on src/reboost/hpge/utils.py:
```diff
@@ -0,0 +1,66 @@
+
+    data = AttrsDict({k: data[k].view_as("np", with_units=True) for k in ("r", "z", field)})
+
+    interpolator = RegularGridInterpolator((data.r.m, data.z.m), data[field].m)
```
what about allowing the user to pass some kwargs to `RegularGridInterpolator` ?

gipert replied:
yeah good idea

willquinn commented:
The way this reader is written implies that we will have 1 drift time map file per detector. Is this what you wish to be the case?

gipert commented:
![image](https://github.com/user-attachments/assets/8cb60e7d-3797-4b0b-af5a-7a7976f429f4)

willquinn commented:
What is the spacing on the grid?

gipert commented:
roughly 1 mm (to generate some map quickly). anyways have a look at the script i have uploaded

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -1,31 +1,30 @@
 
 def r90(edep: ak.Array, xloc: ak.Array, yloc: ak.Array, zloc: ak.Array) -> Array:
-    """Computes R90 for each hit in a ged.
+    """Computes R90 for each hit in an HPGe detector.
```
I dont think the capital in `r90`, I also think the docstring should say a bit what `r90` is.

gipert replied:
i agree but not in this PR

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -1,31 +1,30 @@
 log = logging.getLogger(__name__)
 
 
 def r90(edep: ak.Array, xloc: ak.Array, yloc: ak.Array, zloc: ak.Array) -> Array:
```
I think the function should support reading either `ak.Array` or Array or VoV, and convert in the second case

gipert replied:
can we write a full processor specification somewhere in the docs?

tdixon97 commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +71,178 @@ def cumsum(layout, **_kwargs):
     return Array(ak.flatten(r90).to_numpy())
+
+
+def drift_time(
```
Similar comment to above, support reading in LGDO objects

tdixon97 commented on tests/hpge/simulation/V99999A.yaml:
```diff
@@ -0,0 +1,26 @@
+name: V99999B
```
why dont we get this file from the test data?

gipert replied:
have to check with david if it has the impurity profile in the right format

gipert replied:
I think not because this:

https://github.com/legend-exp/legend-testdata/blob/main/data/legend/metadata/hardware/detectors/germanium/crystals/V99000.yaml

would need to include the `impurity_curve` field? @hervasa2 @fhagemann

hervasa2 replied:
Yes you need to add the field. Copy it from another detector and modify it a bit. If there is no `impurity_curve` field a constant impurity of 0 is set by `LegendDataManagement jl`

gipert replied:
could you or @fhagemann make a quick PR?

gipert replied:
https://github.com/legend-exp/legend-testdata/pull/29

tdixon97 commented on tests/hpge/simulation/geometry.gdml:
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0" ?>
```
why dont we make this file during the tests

gipert replied:
bah not really needed, it's just few lines of gdml. the python script is just there "for courtesy"

tdixon97 commented on tests/hpge/test_dt_heuristic.py:
```diff
@@ -0,0 +1,89 @@
+gamma_stp = ak.Array(
+    {
+        "edep": [
+            [0.0571, 0.00863, 0.026, 0.072, 21.6, 1.09, 485],
```
uuummm what is this?

gipert replied:
this what

tdixon97 replied:
this huge gamma_stp hardcoded akArray

gipert replied:
obviously we need to start having some remage test data somewhere. we can upload something to legend-testdata already or hardcode simple stuff here

tdixon97 replied:
I was planning to do it, now the reshaping is merged

tdixon97 replied:
i can make a PR of a small file there

gipert replied:
yes and then i can update this

tdixon97 commented on tests/hpge/test_dt_heuristic.py:
```diff
@@ -0,0 +1,89 @@
+    )
+
+
+def test_drift_time(dt_map):
```
I think we need more tests also with some similar cases like dt = z

gipert replied:
ok can try something out

ManuelHu replied:
@tdixon97 Do we actually have this feature? Or since when do we get a nice pint-position out of GDML?

tdixon97 replied:
Ok so what we have is to get the position in this form.
```
<class 'pyg4ometry.gdml.Defines.Position'>
```
which i usually call `.eval()` on to get the position as a list. And so I was wrong about it being already in the `pint` format. I see it has a `unit` field, but `eval` gives you the position in another unit (mm).

gipert replied:
@ManuelHu I guess you meant to comment on the coord offset

tdixon97 replied:
```python
pos = [u.parse_expression(f"{getattr(position,coord).eval()}*{position.unit}") for coord in ["x","y","z"]]
```
... pretty ugly but would work

gipert commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +74,158 @@ def cumsum(layout, **_kwargs):
+
+    .. math::
+
+        T_1 = \frac{\sum_{i < m} t_i E_i}{\sum_{i < m} E_i}
```
how... the math seems clear to me

gipert commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +74,158 @@ def cumsum(layout, **_kwargs):
+    yloc: ArrayLike,
+    zloc: ArrayLike,
+    dt_map: HPGeScalarRZField,
+    coord_offset: Sequence[float] = (0, 0, 0),
```
sure, what is the exact type?

gipert commented on src/reboost/hpge/psd.py:
```diff
@@ -72,3 +74,158 @@ def cumsum(layout, **_kwargs):
+
+    Accepts Awkward arrays and uses Numba to speed up the computation.
+
+    For each hit (collection of steps), the drift times and corresponding
```
...isn't that correct? the dt heuristic is a number about the full hit

tdixon97 replied:
ok now I think I understand what you mean
